backend/services/AccountService.ts

```diff
@@ -0,0 +1,219 @@
+      throw e;
+    }
+  }
+
+  /**
+   * Retrieves an account based on the provided reference ID.
+   * @param referenceId - The reference ID used to correlate the data from the blockchain transaction to the account.
+   * @returns A Promise that resolves to an AccountResponse object if the account is found, or undefined if not found.
+   * @throws Throws an error if there was an issue retrieving the account.
+   */
+  public async getAccountByReferenceId(referenceId: string): Promise<AccountResponse | undefined> {
+    // In this case, we're using the referenceId to get the account
+    // Check the webhook and see if the referenceId has been processed
+    logger.debug(`AccountService: getAccountByReferenceId: Looking for account for referenceId:(${referenceId})`);
+    try {
+      const accountData = WebhookController.referenceIdsReceived.get(referenceId);
+      if (accountData) {
+        logger.debug(`Found account for referenceId:(${referenceId})`);
+        return {
+          accessToken: createAuthToken(accountData.accountId),
+          expires: Date.now() + 24 * 60 * 60 * 1_000,
+          referenceId: referenceId,
+          msaId: parseInt(accountData.msaId),
+          displayHandle: accountData.displayHandle,
+        };
+      }
+    } catch (e) {
+      logger.error(`Failed to get account for referenceId:(${referenceId}) error:${e}`);
+      throw e;
+    }
+  }
+
+  /**
+   * Signs in or signs up a user based on the provided request.
+   * If `signUp` is true, it will sign up the user. If `signIn` is true, it will sign in the user.
+   * @param request - The request object containing the necessary information for signing in or signing up.
+   * @returns A promise that resolves to a `WalletLoginResponse` object.
+   * @throws If there is an error during the sign in or sign up process.
+   */
+  public async signInOrSignUp(request: WalletLoginRequestDto): Promise<WalletLoginResponse> {
+    const { signIn, signUp } = request;
+    let response: Partial<WalletLoginResponse> = {};
+    try {
+      if (signUp) {
+        response = await AccountService.getInstance().then((service) => service.signUp(request));
+        return response as WalletLoginResponse;
+      } else if (signIn) {
+        response = await AccountService.getInstance().then((service) => service.signIn(request));
+      }
+      return response as WalletLoginResponse;
+    } catch (e) {
+      logger.error('Failed to sign in or sign up: ', e);
+      throw e;
+    }
+  }
+
+  /**
+   * Signs up a user based on the provided SIWF payload.
+   * @param payload - The SIWF payload containing the signup information.
+   * @returns A Promise that resolves to an object with the referenceId, accessToken, and expires properties.
+   * @throws {HttpError} If the signup payload is invalid or if signup validation fails.
+   */
+  public async signUp(payload: WalletLoginRequestDto): Promise<any> {
+    const api = await getApi();
```

Comment:
nit: this has caused confusion in the past--maybe rename to `chainApi`, to distinguish from a REST api?